In [ ]:
import pandas as pd

In [2]:
import warnings 
warnings.filterwarnings("ignore")

In [3]:
#load data
data=pd.read_csv("creditcard.csv")

# Data Preprocessing

In [4]:
data.head()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0.0,-1.359807,-0.072781,2.536347,1.378155,-0.338321,0.462388,0.239599,0.098698,0.363787,...,-0.018307,0.277838,-0.110474,0.066928,0.128539,-0.189115,0.133558,-0.021053,149.62,0
1,0.0,1.191857,0.266151,0.166480,0.448154,0.060018,-0.082361,-0.078803,0.085102,-0.255425,...,-0.225775,-0.638672,0.101288,-0.339846,0.167170,0.125895,-0.008983,0.014724,2.69,0
2,1.0,-1.358354,-1.340163,1.773209,0.379780,-0.503198,1.800499,0.791461,0.247676,-1.514654,...,0.247998,0.771679,0.909412,-0.689281,-0.327642,-0.139097,-0.055353,-0.059752,378.66,0
3,1.0,-0.966272,-0.185226,1.792993,-0.863291,-0.010309,1.247203,0.237609,0.377436,-1.387024,...,-0.108300,0.005274,-0.190321,-1.175575,0.647376,-0.221929,0.062723,0.061458,123.50,0
4,2.0,-1.158233,0.877737,1.548718,0.403034,-0.407193,0.095921,0.592941,-0.270533,0.817739,...,-0.009431,0.798278,-0.137458,0.141267,-0.206010,0.502292,0.219422,0.215153,69.99,0


In [5]:
data.shape

(284807, 31)

In [6]:
data.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [8]:
data.isna().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [9]:
data.duplicated().sum()

np.int64(1081)

In [10]:
data.drop_duplicates(inplace=True,ignore_index=True)  #ignore index can resets the index of the dataframe

In [ ]:
data.corr()

In [12]:
#breaking our data into X and y
X=data.drop("Class",axis=1)
y=data.Class

In [13]:
y.value_counts()

Class
0    283253
1       473
Name: count, dtype: int64

In [14]:
#Lets we can train-test split our Data
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=123)

## Data Balancing

In [15]:
#Lets we can use the SMOTE
from imblearn.over_sampling import SMOTE
smote=SMOTE(sampling_strategy="minority",random_state=0)

In [16]:
X_train,y_train=smote.fit_resample(X_train,y_train)

In [17]:
X_train.shape

(424890, 30)

In [18]:
y_train.value_counts()

Class
0    212445
1    212445
Name: count, dtype: int64

In [47]:
X_train,X_val,y_train,y_val=train_test_split(X_train,y_train,test_size=0.20,stratify=y_train,random_state=42)  #stratify=y_train means that make the proportion of test(validation) ,according to the proportion of y_train. 

## Apply Normalization and Standardization Scaling On our Data

In [48]:
from sklearn.preprocessing import MinMaxScaler,StandardScaler

In [49]:
#Normalization
mmscaler=MinMaxScaler()
mmscaled_train=mmscaler.fit_transform(X_train)
mmscaled_test=mmscaler.transform(X_test)

In [50]:
#Standardization
stscaler=StandardScaler()
stscaled_train=stscaler.fit_transform(X_train)
stscaled_val=stscaler.transform(X_val)
stscaled_test=stscaler.transform(X_test)

# Applying Our ANN Model

In [51]:
#Importing Libraries
from tensorflow import keras
from keras import layers

In [52]:
# model_ar=keras.Sequential([
#     layers.Input(shape=(X_train.shape[1],)),  #Or (30,)
#     layers.Dense(32,activation="relu"),
#     layers.Dense(16,activation="relu"),
#     layers.Dense(1,activation="sigmoid")

# ])

In [53]:
# model_ar.compile(optimizer="adam",loss="binary_crossentropy",metrics=["accuracy","precision","recall"])

In [54]:
def create_model(activation, optimizer):
    model = keras.Sequential([
        layers.Input((X_train.shape[-1],)),
        layers.Dense(32, activation=activation),
        layers.Dense(16, activation=activation),
        layers.Dense(1, activation="sigmoid")
    ])

    model.compile(
        optimizer=optimizer,
        loss="binary_crossentropy",
        metrics=["accuracy","precision","recall"]
    )

    return model

In [55]:
#model_ar.fit(stscaled_train,y_train,epochs=20,validation_split=0.20,batch_size=128)
#model_ar.fit(stscaled_train,y_train,epochs=50,validation_split=0.20,batch_size=128)
# model_ar.fit(stscaled_train,y_train,epochs=100,validation_split=0.20,batch_size=128)

## Model1 Fitting 

In [56]:
# model1_20 = create_model("relu", "adam")

# model1_20.fit(
#     stscaled_train,
#     y_train,
#     epochs=20,
#     validation_data=(stscaled_val,y_val),
#     batch_size=128
# )

In [57]:
model1_50 = create_model("relu", "adam")

model1_50.fit(
    stscaled_train,
    y_train,
    epochs=50,
    validation_data=(stscaled_val,y_val),
    batch_size=128

)

Epoch 1/50
1700/1700 ━━━━━━━━━━━━━━━━━━━━ 13s 6ms/step - accuracy: 0.9677 - loss: 0.0855 - precision: 0.9799 - recall: 0.9550 - val_accuracy: 0.9879 - val_loss: 0.0339 - val_precision: 0.9892 - val_recall: 0.9867
Epoch 2/50
1700/1700 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9920 - loss: 0.0245 - precision: 0.9931 - recall: 0.9909 - val_accuracy: 0.9950 - val_loss: 0.0185 - val_precision: 0.9915 - val_recall: 0.9985
Epoch 3/50
1700/1700 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.9960 - loss: 0.0142 - precision: 0.9952 - recall: 0.9968 - val_accuracy: 0.9975 - val_loss: 0.0108 - val_precision: 0.9959 - val_recall: 0.9992
Epoch 4/50
1700/1700 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9976 - loss: 0.0095 - precision: 0.9967 - recall: 0.9986 - val_accuracy: 0.9978 - val_loss: 0.0082 - val_precision: 0.9966 - val_recall: 0.9991
Epoch 5/50
1700/1700 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9981 - loss: 0.0076 - precision: 0.9972 - recall: 0.9990 - val_accuracy: 0.9988 - 

## Model2 Fiting 

In [59]:
model2_50 = create_model("tanh", "sgd")

model2_50.fit(
    stscaled_train,
    y_train,
    epochs=50,
    validation_data=(stscaled_val,y_val),
    batch_size=128

)

Epoch 1/50
1700/1700 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step - accuracy: 0.9405 - loss: 0.1660 - precision: 0.9800 - recall: 0.8994 - val_accuracy: 0.9642 - val_loss: 0.1005 - val_precision: 0.9817 - val_recall: 0.9461
Epoch 2/50
1700/1700 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9672 - loss: 0.0912 - precision: 0.9815 - recall: 0.9524 - val_accuracy: 0.9709 - val_loss: 0.0816 - val_precision: 0.9812 - val_recall: 0.9601
Epoch 3/50
1700/1700 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9717 - loss: 0.0782 - precision: 0.9825 - recall: 0.9604 - val_accuracy: 0.9741 - val_loss: 0.0720 - val_precision: 0.9827 - val_recall: 0.9652
Epoch 4/50
1700/1700 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9747 - loss: 0.0700 - precision: 0.9843 - recall: 0.9648 - val_accuracy: 0.9764 - val_loss: 0.0648 - val_precision: 0.9848 - val_recall: 0.9677
Epoch 5/50
1700/1700 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.9769 - loss: 0.0636 - precision: 0.9860 - recall: 0.9677 - val_accuracy: 0.9781 - v

## Model 3 Fitting

In [ ]:
model3_100 = create_model("relu","rmsprop")

model3_100.fit(
    stscaled_train,
    y_train,
    epochs=100,
    validation_data=(stscaled_val,y_val),
    batch_size=128

)

model_ar,adam,batch_size=64

In [61]:
#Epoch 20/20
#5312/5312 ━━━━━━━━━━━━━━━━━━━━ 18s 3ms/step - accuracy: 0.9996 - loss: 0.0013 - precision: 0.9992 - recall: 0.9997 - val_accuracy: 1.0000 - val_loss: 3.5467e-04 - val_precision: 1.0000 - val_recall: 1.0000

model_ar,adam,batch_size=128

In [62]:
#Epoch 20/20
#2656/2656 ━━━━━━━━━━━━━━━━━━━━ 8s 3ms/step - accuracy: 0.9999 - loss: 4.6336e-04 - precision: 0.9998 - recall: 0.9999 - val_accuracy: 1.0000 - val_loss: 8.6220e-05 - val_precision: 1.0000 - val_recall: 1.0000

model_ar,adam,batch_size=256

In [63]:
#Epoch 20/20
#1328/1328 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9999 - loss: 2.3872e-04 - precision: 0.9999 - recall: 0.9999 - val_accuracy: 0.9996 - val_loss: 0.0011 - val_precision: 1.0000 - val_recall: 0.9996

Lets the best results,according to the val_accuracy and others with relation to the training accuracy,The best batch size is 128 for this model.

## Model prediction on unseen(test) Data

#### Predictions For Model 1

In [64]:
# y_pred1_20=model1_20.predict(stscaled_test)
# y_pred1_20=(y_pred1_20 >=0.5).astype(int).ravel()   #astype(int) can converts it to 0 and 1 ,and ravel can convert it into 1D array

#It exactly like this
#y_pred = [1 if prob[0] >= 0.5 else 0 for p in y_pred1_20]
#prob[0] can picks our number at there
# prob       → [0.23]
# prob[0]    → 0.23

In [65]:
y_pred1_50=model1_50.predict(stscaled_test)
y_pred1_50=(y_pred1_50 >=0.7).astype(int).ravel()

# y_prob = model1_50.predict(stscaled_test).ravel()

# for threshold in [0.3, 0.4, 0.5, 0.6, 0.7]:
#     y_pred = (y_prob >= threshold).astype(int)

#     print("Threshold:", threshold)
#     print("Precision:", precision_score(y_test, y_pred, zero_division=0))
#     print("Recall:", recall_score(y_test, y_pred, zero_division=0))
#     print("F1:", f1_score(y_test, y_pred, zero_division=0))
#     print()

2217/2217 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step


#### Predictions For Model 2

In [66]:
y_pred2_50=model2_50.predict(stscaled_test)
y_pred2_50=(y_pred2_50 >= 0.7).astype(int).ravel()

2217/2217 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step


#### Predictions For Model 3

In [67]:
y_pred3_100=model3_100.predict(stscaled_test)
y_pred3_100=(y_pred3_100 >= 0.7).astype(int).ravel()

2217/2217 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step


## Finding confusion_matrix,accuracy_score,precision_score,recall_score,f1_score

In [68]:
from sklearn.metrics import confusion_matrix,accuracy_score,precision_score,recall_score,f1_score

# For Model 1

In [ ]:
#With 20 Epochs
# accuracy1_20 = accuracy_score(y_test, y_pred1_20)
# precision1_20 = precision_score(y_test, y_pred1_20)
# recall1_20 = recall_score(y_test, y_pred1_20)
# f1_1_20 = f1_score(y_test, y_pred1_20)
# confusion_matrix1_20=confusion_matrix(y_test,y_pred1_20)

In [70]:
#With 50 Epochs
accuracy1_50 = accuracy_score(y_test, y_pred1_50)
precision1_50 = precision_score(y_test, y_pred1_50)
recall1_50 = recall_score(y_test, y_pred1_50)
f1_1_50 = f1_score(y_test, y_pred1_50)
confusion_matrix1_50=confusion_matrix(y_test,y_pred1_50)

# For Model 2

In [71]:
#With 50 Epochs
accuracy2_50 = accuracy_score(y_test, y_pred2_50)
precision2_50 = precision_score(y_test, y_pred2_50)
recall2_50 = recall_score(y_test, y_pred2_50)
f1_2_50 = f1_score(y_test, y_pred2_50)
confusion_matrix2_50=confusion_matrix(y_test,y_pred2_50)

# For Model 3

In [72]:
#With 100 Epochs
accuracy3_100 = accuracy_score(y_test, y_pred3_100)
precision3_100 = precision_score(y_test, y_pred3_100)
recall3_100 = recall_score(y_test, y_pred3_100)
f1_3_100 = f1_score(y_test, y_pred3_100)
confusion_matrix3_100=confusion_matrix(y_test,y_pred3_100)

### Models Results:

In [73]:
#Model 1
print(accuracy1_50)
print(precision1_50)
print(recall1_50)
print(f1_1_50)

0.9989567473072802
0.6666666666666666
0.8064516129032258
0.7299270072992701


In [74]:
#model 1
print(confusion_matrix1_50)  #CONFUSION MATRIX

[[70758    50]
 [   24   100]]


In [75]:
#model 2 
print(accuracy2_50)
print(precision2_50)
print(recall2_50)
print(f1_2_50)

0.9972085941465065
0.3724137931034483
0.8709677419354839
0.5217391304347826


In [76]:
#model 2   #CONFUSION MATRIX
print(confusion_matrix2_50)

[[70626   182]
 [   16   108]]


In [77]:
#model 3
print(accuracy3_100)
print(precision3_100)
print(recall3_100)
print(f1_3_100)

0.9984210229515592
0.5306122448979592
0.8387096774193549
0.65


In [78]:
#Model 3    #CONFUSON MATRIX
print(confusion_matrix3_100)

[[70716    92]
 [   20   104]]


## Models Comparasion Chart:

In [81]:
pd.DataFrame({"Model":["Model 1","Model 2","Model 3"],"Activation":["Relu","Tanh","Relu"],"Optimizer":["Adam","SGD","RMSProp"],"Epochs":[50,50,100],"Acuracy":[0.9989,0.9972,0.9984],"Precision":[0.6666,0.3724,0.5306],"Recall":[0.8064,0.8709,0.8387],"F1_score":[0.7299,0.5217,0.65]})

,Model,Activation,Optimizer,Epochs,Acuracy,Precision,Recall,F1_score
0,Model 1,Relu,Adam,50,0.9989,0.6666,0.8064,0.7299
1,Model 2,Tanh,SGD,50,0.9972,0.3724,0.8709,0.5217
2,Model 3,Relu,RMSProp,100,0.9984,0.5306,0.8387,0.6500


# Conclusion:

Among the three ANN models, **Model 1 (ReLU, Adam, 50 epochs)** performed best overall, achieving **99.89% accuracy, 66.66% precision, 80.64% recall, and 72.99% F1-score**. Although Model 2 had higher recall, Model 1 provided the best balance between precision and recall, making it the preferred model.
